### PR TITLE
basic support for running in an user namespace

### PR DESCRIPTION
**Pull request by giuseppe**
**- What I did**
fixed some issues to allow CRI-O running in an user namespace

**- How I did it**
Allow to configure all directories used by CRI-O

**- How to verify it**
create a namespace, and run CRI-O with a configuration file where all directories are writeable by the user

### CODE REVIEW

**Comment by AkihiroSuda** [comment]
:tada: I'll try to add cri-o to Usernetes https://github.com/rootless-containers/usernetes

How did you test rootless cri-o?
With Usernetes or with crictl?

File: server/rootless.go [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+import (
+	"strings"
+
+	rspec "github.com/opencontainers/runtime-spec/specs-go"
+	"github.com/opencontainers/runtime-tools/generate"
+)
+
+func makeOCIConfigurationRootless(g *generate.Generator) error {
+	g.Config.Linux.Resources = nil
+	g.Config.Process.OOMScoreAdj = nil
```
Comment (AkihiroSuda) [inline server/rootless.go]: I'm not familiar with apparmor yet but probably it does not work with rootless?

**Comment by giuseppe** [comment]
@AkihiroSuda I've only used crictl with some basic containers, I'd expect it to fail with more complex ones :-) I am going to test it with Usernetes

File: server/rootless.go (hunk 1 above)
Comment (giuseppe) [reply server/rootless.go]: I am not familiar either but I guess it won't work, similarly with SELinux, I'll force them to ""

**Comment by giuseppe** [comment]
dropping WIP as I could test it with Usernetes and I am able to deploy a simple POD.

File: lib/config.go [hunk 2]
```diff
@@ -96,6 +96,9 @@ type RootConfig struct {
 	// File-based locking is required when multiple users of lib are
 	// present on the same system
 	FileLocking bool `toml:"file_locking"`
+
+	// FileLockingPath specifies the path to use for the locking.
+	FileLockingPath bool `toml:"file_locking_path"`
```
Comment (mrunalp) [inline lib/config.go]: nit: can we fix this commit to not introduce this as bool and rewrite as string in next commit?

**Comment by mrunalp** [comment]
I have one nit, otherwise looks fine :+1:

File: server/rootless.go [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+		}
+		g.Config.Mounts[i].Options = newOptions
+	}
+
+	g.RemoveMount("/sys")
+	sysMnt := rspec.Mount{
+		Destination: "/sys",
+		Type:        "bind",
+		Source:      "/sys",
+		Options:     []string{"nosuid", "noexec", "nodev", "ro", "rbind"},
```
Comment (AkihiroSuda) [inline server/rootless.go]: `ro, rbind` does not make `/sys/fs/cgroup` read-only recursively, could you cherry-pick https://gist.github.com/AkihiroSuda/889b29c56a3c9b546a843fb03f11d60a ?

Comment (giuseppe) [reply server/rootless.go]: I am seeing an issue when I try to mount cgroup after the `/sys/` bind mount.  I am trying to fix it in podman first, as the same bind mount is used there and I am using this patch:

```diff
diff --git a/pkg/spec/spec.go b/pkg/spec/spec.go
index d9888e99..1bcb92ad 100644
--- a/pkg/spec/spec.go
+++ b/pkg/spec/spec.go
@@ -44,6 +44,13 @@ func CreateConfigToOCISpec(config *CreateConfig) (*spec.Spec, error) { //nolint
                        Options:     []string{"nosuid", "noexec", "nodev", "ro", "rbind"},
                }
                g.AddMount(sysMnt)
+               cgroupMnt := spec.Mount{
+                       Destination: "/sys/fs/cgroup",
+                       Type:        "cgroup",
+                       Source:      "cgroup",
+                       Options:     []string{"nosuid", "noexec", "nodev", "relatime", "ro"},
+               }
+               g.AddMount(cgroupMnt)
        }
        if rootless.IsRootless() {
                g.RemoveMount("/dev/pts")
```

The error I see is:
```
container_linux.go:336: starting container process caused "process_linux.go:402: container init caused \"rootfs_linux.go:57: mounting \\\"/sys/fs/cgroup\\\" to rootfs \\\"/tmp/test/rootfs\\\" at \\\"/sys/fs/cgroup\\\" caused \\\"stat /tmp/test/rootfs/sys/fs/cgroup/systemd/user.slice/user-1000.slice/user@1000.service/gnome-terminal-server.service/foo: no such file or directory\\\"\""
```

Comment (giuseppe) [reply server/rootless.go]: Also, is it possible to exploit it without privileged helpers?

Comment (AkihiroSuda) [reply server/rootless.go]: Sorry you need to compile runc with patched `libcontainer/rootfs_linux.go`. I'll open PR in a few days.

As the issue won't be exploited without privileged helpers (e.g. pam_cgfs or manual chown invocation), the fix can be postponed.

File: lib/config.go (hunk 2 above)
Comment (giuseppe) [reply lib/config.go]: absolutely, I thought I already did it, apparently I screwed the rebase.  Pushed a new version

**Comment by giuseppe** [comment]
I've done another change in the rootless code, now `/sys` is mounted from the host only when not creating a new network namespace

File: server/rootless.go (hunk 1 above)
Comment (rhatdan) [reply server/rootless.go]: SELinux would work in UserNS.  SElinux doesn't care about DAC.

**Comment by giuseppe** [comment]
I forgot to address one comment from @rhatdan, I've included this change now:

```diff
@@ -19,7 +19,6 @@ func makeOCIConfigurationRootless(g *generate.Generator) error {
        g.Config.Linux.Resources = nil
        g.Config.Process.OOMScoreAdj = nil
        g.Config.Process.ApparmorProfile = ""
-       g.Config.Process.SelinuxLabel = ""
```

looks like we need to test again :-(

/test all

**Pull request merged**